### ISSUE TITLE
Disable gradient flow in parts of the subgraph

### ISSUE THREAD

**Issue by sai-prasanna** [danijar/ninjax#8]
I am implementing an adversarial loss, where I don't want to compute gradients of the adversary, but want to backprop through the adversary to the generator. Is there an easy way to do this with ninjax? How to make sure that the operation applied by the callable returned by self.get() doesn't track the gradients of the weights that's stored implicitly in the global context when grad is called over this entire thing? 

Edit: I got around doing this, would it be correct?
``` python
class SomeNjModule(nj.Module):
      def loss(self, x):
          # impoure function
          tree_map = jax.tree_util.tree_map
          sg = lambda x: tree_map(jax.lax.stop_gradient, x)
      
          h = self.some_nj_module(x)
          pred = self.adv_nj_module(sg(h))
          prediction_loss = -pred.logprob(gt)
          pure_linear_fn = nj.pure(
              lambda x: self.adv_nj_module(x), nested=True
         )
         adv_params = self.adv_nj_module.getm()
         adv_pred, _ = pure_linear_fn(sg(adv_params), nj.rng(), h)
         adv_loss = -adv_pred.log_prob(
               jnp.zeros_like(adv_pred.mode(), jnp.float32)
          )
          # other_losses will be computed, left out for brevity
          loss = prediction_loss.mean() + adv_loss.mean() + other_losses
          return loss
 
some_module = SomeNjModule(...)
pure_some_module = nj.jit(nj.pure(lambda x: some_module(x)))
loss, state = pure_some_module(some_module.getm(), nj.rng(), inp)
```

1. Would this make sure that when the gradients of adv_loss backpropagates, it will not affect the parameters of the adversary and when prediction_loss part backprops it will only affect the adversary?
2. And can I do nj.pure + nj.jit in the constructor of the `SomeNjModule` and assign the purefunction to some variable and use it without having to construct it inside loss function again?

**Comment by danijar** [comment]
Yes, that works!

You don't need the lambda, `nj.pure(self.adv_nj_module, nested=True)` would work too.

You can also move it into transformation like this (untested), I may add something like this to Ninjax because this has come up a couple of times:

```python3
def modify_state(fn, modifier):
  fn = pure(fn, nested=True)
  def fn2(*args, **kwargs):
    accessed, modified = nj._prerun(fn, *args, **kwargs)
    needed = {k: v for k, v in nj.context().items() if k in accessed}
    needed = modifier(needed)
    state, outs = fn(needed, *args, seed=nj.seed(None, True), **kwargs)
    changes = {k: v for k, v in state.items() if k in modified}
    if context().modify:
      context().update(changes)
    return outs
  return fn2

frozen = lambda fn: modify_state(fn, lambda state: jax.tree_map(jax.lax.stop_gradient, state))
pred = frozen(self.discriminator)(sample)
```

Most of this is for optimizing memory efficiency, a simplified version would be (untested):

```python3
def frozen(fn):
  def fn2(*args, **kwargs):
    state = jax.tree_map(jax.lax.stop_gradient, nj.context())
    state, outs = pure(fn, nested=True)(state, *args, seed=nj.seed(None, True), **kwargs)
    context().modify and context().update(changes)
    return outs
  return fn2

pred = frozen(self.discriminator)(sample)
```

Yet another solution is just to call `nj.grad` twice, once per parameter set you want to differentiate with respect to, then there is no need for freezing parameters:

```python3
class GAN:

  def __init__(self):
    self.discriminator = MLP(..., name='d')
    self.generator = MLP(..., name='g')

  def d_loss(self, data, fake):
    loss_data = -self.discriminator(data).log_prob(ones).mean()
    loss_fake = -self.discriminator(fake).log_prob(zeros).mean()
    return loss_data + loss_fake

  def g_loss(self):
    noise = jax.random.uniform(nj.seed(), shape)
    sample = self.generator(noise)
    loss = -self.discriminator(sample).log_prob(sample).mean()
    return loss, sample

  def train(self, data):
    g_grad, sample = nj.grad(self.g_loss, self.generator, has_aux=True)()
    d_grad = nj.grad(self.d_loss, self.discriminator)(data, sample)
    # ...


gan = GAN(name='model')
params = nj.init(gan.train)
train = jax.jit(nj.pure(gan.train))
for i, data in enumerate(dataset):
  train(params, data, seed=i)
```

Hope that helps!

**Comment by sai-prasanna** [comment]
Thanks, this is insightful. btw Love the meta-classes stuff you pulled off so elegantly in this library. I know metaclasses exist but  I never reach out to it as a tool. Did you use it for something before?
